pkg/vm/vm.go

```diff
@@ -22,37 +28,302 @@ type VirtualMachineConfig struct {
+	}
+
+	instruction, err := DecodeInstruction(bytecodeInstruction)
+	if err != nil {
+		return fmt.Errorf("cannot decode step at %d: %w", vm.Context.Pc, err)
+	}
+
+	err = vm.RunInstruction(instruction)
+	if err != nil {
+		return fmt.Errorf("cannot run step at %d: %w", vm.Context.Pc, err)
+	}
+
+	vm.Step += 1
 	return nil
 }
+func (vm *VirtualMachine) RunStepAt(pc uint64) error {
+	vm.Context.Pc = pc
+	return vm.RunStep()
+}
 
 func (vm *VirtualMachine) RunInstruction(instruction *Instruction) error {
+	// todo(rodro): any OffOpX can be negative, a better math system is required due to
+	// substraction. Also it will need to handle overflows and underflows
+	dstCell, err := vm.getCellDst(instruction)
+	if err != nil {
+		return err
+	}
+
+	op0Cell, err := vm.getCellOp0(instruction)
+	if err != nil {
+		return err
+	}
+
+	op1Cell, err := vm.getCellOp1(instruction, op0Cell)
+	if err != nil {
+		return err
+	}
+
+	res, err := vm.computeRes(instruction, op0Cell, op1Cell)
+	if err != nil {
+		return err
+	}
+
+	err = vm.opcodeAssertions(instruction, dstCell, op0Cell, res)
+	if err != nil {
+		return err
+	}
+
+	nextPc, err := vm.updatePc(instruction, dstCell, op1Cell, res)
+	if err != nil {
+		return err
+	}
+
+	nextAp, err := vm.updateAp(instruction, res)
+	if err != nil {
+		return err
+	}
+
+	nextFp, err := vm.updateFp(instruction, res)
+	if err != nil {
+		return err
+	}
+
+	vm.Context.Pc = nextPc
+	vm.Context.Ap = nextAp
+	vm.Context.Fp = nextFp
+
+	return nil
+}
+
+func (vm *VirtualMachine) RunHint() error {
 	return nil
 }
 
-func (vm *VirtualMachine) RunHint(instruction *Instruction) error {
+func (vm *VirtualMachine) getCellDst(instruction *Instruction) (*mem.Cell, error) {
+	var dstRegister uint64
+	if instruction.DstRegister == Ap {
+		dstRegister = vm.Context.Ap
+	} else {
+		dstRegister = vm.Context.Fp
+	}
+
+	// todo(rodro): fix this math
+	return vm.MemoryManager.Memory.Peek(dataSegment, dstRegister+uint64(instruction.OffDest))
+}
+
+func (vm *VirtualMachine) getCellOp0(instruction *Instruction) (*mem.Cell, error) {
+	var op0Register uint64
+	if instruction.Op0Register == Ap {
+		op0Register = vm.Context.Ap
+	} else {
+		op0Register = vm.Context.Fp
+	}
+	// todo(rodro): fix this math
+	offset := op0Register + uint64(instruction.OffOp0)
+	return vm.MemoryManager.Memory.Peek(dataSegment, offset)
+}
+
+func (vm *VirtualMachine) getCellOp1(instruction *Instruction, op0Cell *mem.Cell) (*mem.Cell, error) {
+	var op1Address *mem.MemoryAddress
+	switch instruction.Op1Source {
+	case Op0:
+		// in this case Op0 is being used as an address, and must be of unwrapped as is
+		op0Address, err := op0Cell.Read().ToMemoryAddress()
+		if err != nil {
+			return nil, fmt.Errorf("expected op0 to be an address: %w", err)
+		}
+		op1Address = mem.CreateMemoryAddress(op0Address.SegmentIndex, op0Address.Offset)
+	case Imm:
+		// todo(rodro): would it be sensitive to check instruction.OffOp1 == 1?
```

Comment:
Practically speaking I've never seen an instruction that has a weird imm offset and the compiler doesn't spit out those but the semantics allows that, meaning in the future that may be used for some reason

Reply:
ok, so let's leave it checkless since the state transition specs doesn't impose this restriction either